File: Hackerrank/Cycle Detection/CycleDetection.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // driver code already given on hackerrank platform
    static class SinglyLinkedListNode {
        public int data;
        public SinglyLinkedListNode next;

        public SinglyLinkedListNode(int nodeData) {
            this.data = nodeData;
            this.next = null;
        }
    }

    static class SinglyLinkedList {
        public SinglyLinkedListNode head;
        public SinglyLinkedListNode tail;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
        }

        public void insertNode(int nodeData) {
            SinglyLinkedListNode node = new SinglyLinkedListNode(nodeData);

            if (this.head == null) {
                this.head = node;
            } else {
                this.tail.next = node;
            }

            this.tail = node;
        }
    }

    public static void printSinglyLinkedList(SinglyLinkedListNode node, String sep, BufferedWriter bufferedWriter) throws IOException {
        while (node != null) {
            bufferedWriter.write(String.valueOf(node.data));

            node = node.next;

            if (node != null) {
                bufferedWriter.write(sep);
            }
        }
    }

    // Complete the hasCycle function below.

    /*
     * For your reference:
     *
     * SinglyLinkedListNode {
     *     int data;
     *     SinglyLinkedListNode next;
     * }
     *
     */

    static boolean hasCycle(SinglyLinkedListNode head) {

        // using two pointer method : slow pointer(ptr1) & fast pointer(ptr2)
        SinglyLinkedListNode ptr1=head;
        SinglyLinkedListNode ptr2=head;

        // if at some point my slow pointer and fast pointer becomes the same , we can say cycle exists else the cycle don't exist
        while(ptr1!=null && ptr2!=null && ptr1.next!=null){
            ptr1=ptr1.next.next;
            ptr2=ptr2.next;
            if(ptr1==ptr2){
                return true;
            }
        }
        return false;
    }

    private static final Scanner scanner = new Scanner(System.in);

File: Hackerrank/Cycle Detection/Cycle_detection.cpp
// * ALGORITHM ==>
// * In this problem we use two pointer methods , i.e. slow & fast pointer.
// * We keep on traversing the nodes till fast becomes equal to null or fast points to the last node or fast and slow point to the same node.
// * And to detect the cycle, we wait till fast becomes equal to slow, as this can only happen when there is a cycle.
// * Had there not been the cycle, fast never would have been equal to slow and would reach the NULL or last node depending on the size of the Linked List.

#include <bits/stdc++.h>

using namespace std;

class SinglyLinkedListNode
{
public:
    int data;
    SinglyLinkedListNode *next;

    SinglyLinkedListNode(int node_data)
    {
        this->data = node_data;
        this->next = nullptr;
    }
};

class SinglyLinkedList
{
public:
    SinglyLinkedListNode *head;
    SinglyLinkedListNode *tail;

    SinglyLinkedList()
    {
        this->head = nullptr;
        this->tail = nullptr;
    }

    void insert_node(int node_data)
    {
        SinglyLinkedListNode *node = new SinglyLinkedListNode(node_data);

        if (!this->head)
        {
            this->head = node;
        }
        else
        {
            this->tail->next = node;
        }

        this->tail = node;
    }
};

void print_singly_linked_list(SinglyLinkedListNode *node, string sep, ofstream &fout)
{
    while (node)
    {
        fout << node->data;

        node = node->next;

        if (node)
        {
            fout << sep;
        }
    }
}

void free_singly_linked_list(SinglyLinkedListNode *node)
{
    while (node)
    {
        SinglyLinkedListNode *temp = node;
        node = node->next;

        free(temp);
    }
}

/*
 * For your reference:
 *
 * SinglyLinkedListNode {
 *     int data;
 *     SinglyLinkedListNode* next;
 * };
 *
 */
// Complete the has_cycle function below.

bool has_cycle(SinglyLinkedListNode *head)
{
    if (head == NULL)
    {
        return false;
    }
    SinglyLinkedListNode *slow = head, *fast = head;
    bool cycle = false;
    while (fast != NULL && fast->next != NULL)
    {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
        {
            cycle = true;
            break;
        }
    }
    return cycle;
}

// driver code
int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int tests;
    cin >> tests;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    for (int tests_itr = 0; tests_itr < tests; tests_itr++)
    {
        int index;
        cin >> index;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');

        SinglyLinkedList *llist = new SinglyLinkedList();

        int llist_count;
        cin >> llist_count;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');

        for (int i = 0; i < llist_count; i++)
        {
            int llist_item;
            cin >> llist_item;
            cin.ignore(numeric_limits<streamsize>::max(), '\n');

            llist->insert_node(llist_item);
        }

        SinglyLinkedListNode *extra = new SinglyLinkedListNode(-1);
        SinglyLinkedListNode *temp = llist->head;

        for (int i = 0; i < llist_count; i++)
        {
            if (i == index)
            {
                extra = temp;
            }

            if (i != llist_count - 1)
            {
                temp = temp->next;
            }
        }

        temp->next = extra;

        bool result = has_cycle(llist->head);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}
